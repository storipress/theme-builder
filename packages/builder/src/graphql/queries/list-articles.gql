query ListArticles($page: Int! = 1, $desk_ids: [ID!], $first: Int! = 30, $featured: Boolean) {
  articles(
    page: $page
    desk_ids: $desk_ids
    first: $first
    published: true
    featured: $featured
    sortBy: [{ column: PUBLISHED_AT, order: DESC }]
  ) {
    paginatorInfo {
      hasMorePages
    }
    data {
      id
      title
      desk {
        id
        name
        slug
        desk {
          id
          name
          slug
        }
      }
      featured
      slug
      blurb
      cover
      created_at
      document
      published
      published_at
      stage {
        ready
      }
      seo
      authors {
        id
        name: full_name
      }
    }
  }
}
