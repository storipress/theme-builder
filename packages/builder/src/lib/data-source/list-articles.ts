import type { Desk, ResolvedArticle } from './types'
import { Array, Effect, pipe } from 'effect'
import pMemoize from 'p-memoize'
import type { Article, ListArticlesVariables } from '~/api'
import { getSite, listArticles } from '~/api'
import type { ListArticlesQuery } from '~/generated/graphql'
import { store } from '~/store'
import { transformArticles } from './transform-article'

export const canLoadMore = ref(true)

export const memoizedListArticles = pMemoize(async (listArticlesVariables: ListArticlesVariables) => {
  return listArticles(listArticlesVariables)
})

const EACH_PAGE_COUNT = 50

function listArticles$(input: {
  page: number
  first: number
  featured: boolean
  deskIds?: string[]
}): Effect.Effect<ListArticlesQuery['articles']['data']> {
  return pipe(
    Effect.promise(() => memoizedListArticles(input)),
    Effect.tap(({ paginatorInfo }: ListArticlesQuery['articles']) => {
      if (paginatorInfo.hasMorePages) {
        // If there has more article, we will set our flag
        canLoadMore.value = true
      }
    }),
    Effect.map(({ data }) => data),
  )
}

export async function doListArticlesToPage(page: number, desks: Desk[]): Promise<Article[]> {
  const deskIds = pipe(
    desks,
    Array.map((desk) => [desk.id, ...Array.map(desk.desks, (d) => d.id)]),
  )

  // Assume there is no more
  canLoadMore.value = false

  // Safety: has memoized list articles API
  const fetchWhenDeskEmpty = (index: number) => [
    listArticles$({
      page: index + 1,
      first: EACH_PAGE_COUNT,
      featured: false,
    }),
    listArticles$({ page: index + 1, first: EACH_PAGE_COUNT, featured: true }),
  ]

  // will fetch 50 article per desks for use
  const fetchWhenDeskNotEmpty = (index: number) =>
    pipe(
      deskIds,
      Array.map((deskIds) =>
        listArticles$({
          page: index + 1,
          deskIds,
          first: EACH_PAGE_COUNT,
          featured: false,
        }),
      ),
      Array.append(listArticles$({ page: index + 1, first: EACH_PAGE_COUNT, featured: true })),
    )

  const pages = await pipe(
    Array.makeBy(page, deskIds.length > 0 ? fetchWhenDeskNotEmpty : fetchWhenDeskEmpty),
    Array.flatten,
    Effect.allWith({ concurrency: 'unbounded' }),
    Effect.map(Array.flatten),
    Effect.runPromise,
  )
  return pages
}

export function listArticlesToPage(page: number): Promise<Article[]> {
  const storeState = store.state as typeof store.state & { front: { desks: Desk[] } }
  const desks = storeState.front.desks
  return doListArticlesToPage(page, desks)
}

export const getArticlesMap = pMemoize(async (_clientId: string, page: number): Promise<ResolvedArticle[]> => {
  const [site, raw] = await Promise.all([getSite(), listArticlesToPage(page)])

  return transformArticles(raw, site)
})
