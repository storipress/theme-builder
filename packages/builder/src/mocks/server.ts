import type { RequestHandler } from 'msw'
import { pipe, Record } from 'effect'
import { setupServer } from 'msw/node'

const handlers = pipe(
  import.meta.glob<{ handler: RequestHandler }>('./graphql/*.ts', { eager: true }),
  Record.map((m) => m.handler),
  Record.values,
)

export const server = setupServer(...handlers)
