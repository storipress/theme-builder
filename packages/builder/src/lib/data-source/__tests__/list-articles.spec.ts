import type { Desk } from '../types'
import { expect, it } from 'vitest'
import { server } from '~/mocks/server'
import { doListArticlesToPage } from '../list-articles'

server.listen()

it('can list articles', async () => {
  const desks = [
    {
      id: '1',
      name: 'desk1',
      desks: [],
    },
    {
      id: '2',
      name: 'desk2',
      desks: [
        {
          id: '4',
        },
        {
          id: '5',
        },
      ],
    },
    {
      id: '3',
      name: 'desk3',
      desks: [
        {
          id: '6',
        },
      ],
    },
  ]
  const articles = await doListArticlesToPage(1, desks as Desk[])

  expect(articles).toMatchSnapshot()
  // mock article length * 2, because we will fetch featured article which may cause duplicate
  expect(articles).toHaveLength(20)
})
