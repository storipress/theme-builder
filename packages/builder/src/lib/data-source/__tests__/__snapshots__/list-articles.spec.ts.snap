// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`can list articles 1`] = `
[
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "1",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "1",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "1",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "7",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "4",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "2",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "4",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "4",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "5",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "5",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "4",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "8",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "4",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "10",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "6",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "3",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "6",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "6",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "6",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "9",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "1",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "1",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "4",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "2",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "6",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "3",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "4",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "4",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "5",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "5",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "6",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "6",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "1",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "7",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "4",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "8",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "6",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "9",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
  {
    "authors": [],
    "blurb": "",
    "cover": null,
    "created_at": "2022-01-01",
    "desk": {
      "desk": null,
      "id": "4",
      "name": "desk",
      "slug": "desk",
    },
    "document": null,
    "featured": false,
    "id": "10",
    "published": true,
    "published_at": "2020-01-01",
    "seo": null,
    "slug": "slug",
    "stage": {
      "ready": true,
    },
    "title": "title",
  },
]
`;
